CSHARP-2427: Implement Connection Monitoring and Pooling.

DmitryLukyanov opened the pull request:
Evergreen:  https://evergreen.mongodb.com/version/5d8a9762e3c3311042969dbd

DmitryLukyanov commented on tests/MongoDB.Driver.Core.Tests/Specifications/connection-monitoring-and-pooling/ConnectionMonitoringAndPoolingTestRunner.cs:
```diff
@@ -0,0 +1,609 @@
+    public class ConnectionMonitoringAndPoolingTestRunner
+    {
+        #region static
+        private static readonly string[] __permanentIgnoredEvents =
```
This list contains the list of events which should be checked in the tests, since they are not expected by spec. Mostly it's `-ing` events.

DmitryLukyanov replied:
As I see, my comments below are mostly related to the fact that spec doesn't expect any `-ing` events (except of `ConnectionCheckOutStarted`). So, our implementation provides more events which can cause some naming problems(described in the next 2 comments) for events.

rstam replied:
Rename to `__alwaysIgnoredEvents`

DmitryLukyanov commented on tests/MongoDB.Driver.Core.Tests/Specifications/connection-monitoring-and-pooling/ConnectionMonitoringAndPoolingTestRunner.cs:
```diff
@@ -0,0 +1,609 @@
+        #region static
+        private static readonly string[] __permanentIgnoredEvents =
+        {
+            nameof(ConnectionPoolOpenedEvent),
```
I've specified `ConnectionPoolOpenedEvent` ( with `-ed`) event here as not used and at the same time, the test runner expects that `ConnectionPoolOpeningEvent` (with `-ing`) will be considered as `ConnectionPoolCreated` (line 410).

The reason for it in this line: https://github.com/mongodb/mongo-csharp-driver/blob/master/src/MongoDB.Driver.Core/Core/ConnectionPools/ExclusiveConnectionPool.cs#L207.

In this line we create connections. So, we create them after `ConnectionPoolOpeningEvent` and before `ConnectionPoolOpenedEvent`. 

Of cause I can change this code. For example, I can move `ConnectionPoolOpenedEvent` to the line higher than `207` (in `ExclusiveConnectionPool`) or create a new event with exactly `ConnectionPoolCreated` name. But not sure that it worth it.

From my perspective, the current events are good enough. But I can be wrong.

cc: @jyemin , @rstam

rstam replied:
See notes elsewhere about not mapping `-ing` driver events to `-ed` CMAP events.

DmitryLukyanov commented on tests/MongoDB.Driver.Core.Tests/Specifications/connection-monitoring-and-pooling/ConnectionMonitoringAndPoolingTestRunner.cs:
```diff
@@ -0,0 +1,609 @@
+        {
+            nameof(ConnectionPoolOpenedEvent),
+            nameof(ConnectionPoolAddingConnectionEvent),
+            nameof(ConnectionPoolCheckedInConnectionEvent),
```
Similar situation as above.
I've skipped `ConnectionPoolCheckedInConnectionEvent` (with `-ed`) and used `ConnectionPoolCheckingInConnectionEvent` (with `-ing`) to map on `ConnectionCheckedIn`, because of this line: https://github.com/DmitryLukyanov/mongo-csharp-driver/blob/csharp2427/src/MongoDB.Driver.Core/Core/ConnectionPools/ExclusiveConnectionPool.cs#L355 (and `359`).

In these lines we close a connection (via calling `Dispose`) before the time when we call `ConnectionPoolCheckedInConnectionEvent` (`-ed`), but after `ConnectionPoolCheckingInConnectionEvent` (`-ing`). Since several tests expect that we need to have a `checkedIn` event before a `connectionClosed`, I mapped `ConnectionCheckedIn` on `ConnectionPoolCheckingInConnectionEvent`

Again, I can be wrong, but I guess these events are good enough too.

cc: @jyemin , @rstam

rstam replied:
See notes elsewhere.

DmitryLukyanov commented on tests/MongoDB.Driver.Core.Tests/Specifications/connection-monitoring-and-pooling/ConnectionMonitoringAndPoolingTestRunner.cs:
```diff
@@ -0,0 +1,609 @@
+                    connectionPoolSettings.MaxConnections.Should().Be(expectedMaxPoolSize);
+                    var expectedMinPoolSize = expectedOption["minPoolSize"].ToInt32();
+                    connectionPoolSettings.MinConnections.Should().Be(expectedMinPoolSize);
+                    // maxIdleTimeMS : todo:
```
as I see this setting is located in `connection settings`, not in `connectionPool settings`. See this line: https://github.com/DmitryLukyanov/mongo-csharp-driver/pull/57/files#diff-a638783da40724c2c11679e9321e1829R460
So, a connection pool event doesn't know anything about this setting. That's why it's not implemented here. I suggest just skip this option in asserting.
If so, I will remove this `todo` line.

cc: @rstam , @jyemin

rstam replied:
Sure... don't see how you can do anything else but ignore `expectedMaxIdleTimeMS` here...

DmitryLukyanov replied:
done

DmitryLukyanov commented on tests/MongoDB.Driver.Core.Tests/Specifications/connection-monitoring-and-pooling/ConnectionMonitoringAndPoolingTestRunner.cs:
```diff
@@ -0,0 +1,609 @@
+
+            if (expectedEvent.TryGetValue("reason", out var reason))
+            {
+                if (!(actualEvent is ConnectionClosedEvent)) // we don't support a reason for ConnectionClosedEvent
```
as I wrote in the slack, we cannot determinate from the connection inside why a `Dispose` method has been called.
Since `ConnectionClosedEvent` is triggered from the `Dispose`, we cannot validate this condition for this event

DmitryLukyanov commented on tests/MongoDB.Driver.Core.Tests/Specifications/connection-monitoring-and-pooling/ConnectionMonitoringAndPoolingTestRunner.cs:
```diff
@@ -0,0 +1,609 @@
+                }
+                else
+                {
+                    if (tasks[target] != null)
```
`Start` step doesn't start a thread. The only thing which is happened that we reserve a place for the thread in `tasks` collection. Here we just validate that `Start` has been called before this step.

DmitryLukyanov commented on tests/MongoDB.Driver.Core.Tests/Specifications/connection-monitoring-and-pooling/ConnectionMonitoringAndPoolingTestRunner.cs:
```diff
@@ -0,0 +1,609 @@
+            return expectedEvents;
+        }
+
+        private List<object> GetFilteredEvents(EventCapturer eventCapturer, BsonDocument test)
```
this method returns the list of events from `EventCapturer` which we can validate

DmitryLukyanov commented on tests/MongoDB.Driver.Core.Tests/Specifications/connection-monitoring-and-pooling/ConnectionMonitoringAndPoolingTestRunner.cs:
```diff
@@ -0,0 +1,609 @@
+            return eventCapturer.Events.Where(c => ignoredEvents.Contains(c.GetType().Name.ToString()) != true).ToList();
+        }
+
+        private string MapErrorTypeToExpected(Exception exception, out string expectedErrorMessage)
```
Map an error/error message from the spec on an error/error message which we have in the driver

DmitryLukyanov commented on tests/MongoDB.Driver.Core.Tests/Specifications/connection-monitoring-and-pooling/ConnectionMonitoringAndPoolingTestRunner.cs:
```diff
@@ -0,0 +1,609 @@
+                .Returns(() =>
+                {
+                    var connection = new MockConnection(serverId, connectionSettings, eventSubscriber);
+                    connection.Open(CancellationToken.None);
```
this method is required to correctly calculate expire dates

DmitryLukyanov commented on tests/MongoDB.Driver.Core.Tests/Specifications/connection-monitoring-and-pooling/ConnectionMonitoringAndPoolingTestRunner.cs:
```diff
@@ -0,0 +1,609 @@
+        private void Start(BsonDocument operation, ConcurrentDictionary<string, Task> tasks)
+        {
+            var startTarget = operation.GetValue("target").ToString();
+            tasks.GetOrAdd(startTarget, (Task)null);
```
as I see from the spec tests, `start` step is more like a flag that some logic is going to be executed in the thread rather than a real thread. So, here I just mark that a current target will be used as a thread in the next steps.

DmitryLukyanov commented on tests/MongoDB.Driver.Core.Tests/Specifications/connection-monitoring-and-pooling/ConnectionMonitoringAndPoolingTestRunner.cs:
```diff
@@ -0,0 +1,609 @@
+
+    internal static class IdGeneratorReflector
+    {
+        public static void __lastId(object value) => Reflector.SetStaticField(typeof(IdGenerator<ConnectionId>), nameof(__lastId), value);
```
Reset `global` counter for `connectionId`

rstam replied:
Can be more type safe:

    __lastId(int value)

DmitryLukyanov commented on src/MongoDB.Driver.Core/Core/ConnectionPools/ExclusiveConnectionPool.cs:
```diff
@@ -140,9 +144,9 @@ public int UsedCount
-
             var helper = new AcquireConnectionHelper(this);
+
+            ThrowIfNotOpen();
```
I know that it looks not really as we usually do (Having `ThrowIfNotOpen` as not the first checking for a method),
but this test: `pool-checkout-error-closed.json` expects a `ConnectionCheckOutStarted` event before a `ConnectionCheckOutFailed`.
Currently, I trigger `ConnectionCheckOutStarted` in the `AcquireConnectionHelper` constructor and then `ConnectionCheckOutFailed` is thrown in the `ThrowIfNotOpen` method.

Can be discussed.

DmitryLukyanov replied:
>in the AcquireConnectionHelper constructor

also, maybe `started` logging can be moved from the constructor here into a new method in the helper

rstam replied:
See note below.

DmitryLukyanov commented on src/MongoDB.Driver.Core/Core/ConnectionPools/ExclusiveConnectionPool.cs:
```diff
@@ -185,7 +189,12 @@ public async Task<IConnectionHandle> AcquireConnectionAsync(CancellationToken ca
         {
             ThrowIfNotOpen();
+
+            _clearingEventHandler?.Invoke(new ConnectionPoolClearingEvent(_serverId, _settings));
```
new clean events

rstam replied:
Ack

DmitryLukyanov commented on src/MongoDB.Driver.Core/Core/ConnectionPools/ExclusiveConnectionPool.cs:
```diff
@@ -335,20 +344,21 @@ private async Task EnsureMinSizeAsync(CancellationToken cancellationToken)
             var stopwatch = Stopwatch.StartNew();
-            _connectionHolder.Return(connection);
-            _poolQueue.Release();
+            if (_state.Value == State.Disposed)
```
this test `pool-checkin-destroy-closed.json` expects that even if we dispose an already disposed connection, we still need to trigger events `ConnectionPoolCheckingInConnectionEvent` and `ConnectionPoolCheckedInConnectionEvent`

rstam replied:
Replace lines 353-367 with:

    bool closeConnection;
    if (_state.Value == State.Disposed)
    {
        closeConnection = true;
    }
    else
    {
        _connectionHolder.Return(connection, out closeConnection);
        _poolQueue.Release();
    }
    stopwatch.Stop();

    if (_checkedInConnectionEventHandler != null)
    {
        _checkedInConnectionEventHandler(new ConnectionPoolCheckedInConnectionEvent(connection.ConnectionId, stopwatch.Elapsed, EventContext.OperationId));
    }

    if (closeConnection)
    {
        connection.Dispose();
    }

The purpose is to delay closing the connection until after the `ConnectionPoolCheckedInConnectionEvent` has been raised so that events are raised in the right order.

DmitryLukyanov replied:
Still under discussion.
The current implementation has fewer changes and looks more similar to the previously implemented.

DmitryLukyanov commented on tests/MongoDB.Driver.Core.TestHelpers/EventCapturer.cs:
```diff
@@ -19,16 +19,16 @@
-        private readonly IEventSubscriber _subscriber;
         private readonly Dictionary<Type, Func<object, bool>> _eventsToCapture;
+        private readonly IEventSubscriber _subscriber;
+        internal readonly object _lock = new object();
```
this change (together with the line `27` and `80`) is required for a new class `EventNotifier`. 

I don't really like the changes in these lines, but it's the best which I could imagine for now. I will try to think about these lines tomorrow once again

DmitryLukyanov commented on tests/MongoDB.Driver.Core.TestHelpers/EventNotifier.cs:
```diff
@@ -0,0 +1,62 @@
+
+namespace MongoDB.Driver.Core.TestHelpers
+{
+    public class EventNotifier
```
this class is created for the target to get an event(`_taskCompletionSource`) when a specified condition is reached in the `EventCapturer`

DmitryLukyanov commented on tests/MongoDB.Driver.Core.Tests/Core/Helpers/MockConnection.cs:
```diff
@@ -31,21 +32,38 @@ public class MockConnection : IConnection
     {
         // fields
         private ConnectionId _connectionId;
+        private readonly ConnectionSettings _connectionSettings;
```
added several changes to correspond to the particular parts of `BinaryConnection` (`closed/-ing` events and `expire` logic).

rstam replied:
Ack

DmitryLukyanov commented on tests/MongoDB.Driver.Core.Tests/Specifications/connection-monitoring-and-pooling/ConnectionMonitoringAndPoolingTestRunner.cs:
```diff
@@ -0,0 +1,609 @@
+            }
+        }
+
+        private Task CreateTask(Action action)
```
I've tried to use `Thread`s instead of `Task`s for this ticket. But without big results. I'm not sure how I can implement this logic: https://github.com/DmitryLukyanov/mongo-csharp-driver/pull/57/files?file-filters%5B%5D=.cs&file-filters%5B%5D=.json&file-filters%5B%5D=.yml#diff-a638783da40724c2c11679e9321e1829R519

Also, the current implementation based on `tasks` looks good enough for me. I'm open to criticism/questions :)

rstam replied:
I'm fine with `Task`

DmitryLukyanov commented on tests/MongoDB.Driver.Core.Tests/Specifications/connection-monitoring-and-pooling/ConnectionMonitoringAndPoolingTestRunner.cs:
```diff
@@ -0,0 +1,609 @@
+            var waitThread = operation.GetValue("target").ToString();
+            if (tasks.TryGetValue(waitThread, out var task) && task != null)
+            {
+                exception = Record.Exception(() => task.GetAwaiter().GetResult());
```
I use `GetAwaiter().GetResult()` intead of `Wait()` because `Wait()` packs errors in `AggregateException`

DmitryLukyanov commented on tests/MongoDB.Driver.Core.Tests/Specifications/connection-monitoring-and-pooling/ConnectionMonitoringAndPoolingTestRunner.cs:
```diff
@@ -0,0 +1,609 @@
+            }
+        }
+
+        private class ThreadPerTaskScheduler : TaskScheduler
```
this code is copied from that place: https://code.msdn.microsoft.com/Samples-for-Parallel-b4b76364/sourcecode?fileId=44488&pathId=2098696067

As I see from the description and my local tests, this is exactly what we want to see in this test. With this `taskScheduler` each task will be launched in a new thread.

DmitryLukyanov commented on tests/MongoDB.Driver.Core.TestHelpers/EventCapturer.cs:
```diff
@@ -244,5 +254,40 @@ private BsonJavaScriptWithScope RecursivelyMaterialize(BsonJavaScriptWithScope v
             }
         }
+
+        private class EventNotifier
```
Probably this class looks better as a nested rather than a separate as it was implemented in the previous commit.

rstam replied:
This is a really nice class. Clever idea!

vincentkam replied:
I definitely found this class very intriguing:  I especially like the the use of `TaskCompletionSource`  😺

rstam commented on src/MongoDB.Driver.Core/Core/ConnectionPools/ExclusiveConnectionPool.cs:
```diff
@@ -54,6 +54,8 @@ internal sealed class ExclusiveConnectionPool : IConnectionPool
         private readonly Action<ConnectionPoolClosingEvent> _closingEventHandler;
         private readonly Action<ConnectionPoolClosedEvent> _closedEventHandler;
+        private readonly Action<ConnectionPoolClearingEvent> _clearingEventHandler;
+        private readonly Action<ConnectionPoolClearedEvent> _clearedEventHandler;
```
Add:

    private readonly Action<ConnectionCreatedEvent> _connectionCreatedEventHandler;

DmitryLukyanov replied:
done

rstam commented on src/MongoDB.Driver.Core/Core/ConnectionPools/ExclusiveConnectionPool.cs:
```diff
@@ -88,6 +90,8 @@ internal sealed class ExclusiveConnectionPool : IConnectionPool
             eventSubscriber.TryGetEventHandler(out _addingConnectionEventHandler);
             eventSubscriber.TryGetEventHandler(out _addedConnectionEventHandler);
+            eventSubscriber.TryGetEventHandler(out _clearingEventHandler);
+            eventSubscriber.TryGetEventHandler(out _clearedEventHandler);
```
Add:

    eventSubscriber.TryGetEventHandler(out _connectionCreatedEventHandler);

DmitryLukyanov replied:
done

rstam commented on src/MongoDB.Driver.Core/Core/ConnectionPools/ExclusiveConnectionPool.cs:
```diff
@@ -140,9 +144,9 @@ public int UsedCount
+            ThrowIfNotOpen();
             try
             {
                 helper.CheckingOutConnection();
```
Let's split `CheckingOutConnection` into two methods and call `ThrowIfNotOpen` between the two:

    helper.CheckingOutConnection();
    ThrowIfNotOpen();
    helper.EnterWaitQueue();

DmitryLukyanov replied:
done

rstam commented on src/MongoDB.Driver.Core/Core/ConnectionPools/ExclusiveConnectionPool.cs:
```diff
@@ -162,9 +166,9 @@ public IConnectionHandle AcquireConnection(CancellationToken cancellationToken)
-
             var helper = new AcquireConnectionHelper(this);
+
+            ThrowIfNotOpen();
```
See note above.

rstam commented on src/MongoDB.Driver.Core/Core/ConnectionPools/ExclusiveConnectionPool.cs:
```diff
@@ -185,7 +189,12 @@ public async Task<IConnectionHandle> AcquireConnectionAsync(CancellationToken ca
+
             Interlocked.Increment(ref _generation);
+
+            _clearedEventHandler?.Invoke(new ConnectionPoolClearedEvent(_serverId, _settings));
```
This note is for line 203:

Raise the new `ConnectionCreatedEvent` from`CreateNewConnection`:

    var pooledConnection = new PooledConnection(this, connection);
    _connectionCreatedEventHandler?.Invoke(new ConnectionCreatedEvent(connection.ConnectionId, connection.Settings, EventContext.OperationId));
    return pooledConnection;

rstam replied:
This note is for line 216:

Move call to `MaintainSizeAsync` from line 216 to between lines 221-222.

The purpose is to raise `ConnectionPoolOpenedEvent` BEFORE new connections start getting added to the pool to satisfy minPoolSize.

rstam replied:
This note is for lines 320-326. Replace them with:

    _connectionHolder.Return(connection, out var closeConnection); // we expect closeConnection to be set to false
    stopwatch.Stop();

    if (closeConnection)
    {
        connection.Dispose();
    }
    else
    {
        if (_addedConnectionEventHandler != null)
        {
            _addedConnectionEventHandler(new ConnectionPoolAddedConnectionEvent(connection.ConnectionId, stopwatch
        }
    }

DmitryLukyanov replied:
Still under discussion.
The current implementation has fewer changes and looks more similar to the previously implemented.

rstam commented on src/MongoDB.Driver.Core/Core/ConnectionPools/ExclusiveConnectionPool.cs:
```diff
@@ -357,11 +367,21 @@ private void ReleaseConnection(PooledConnection connection)
             }
         }
 
+        private void ThrowAndLogException(Exception ex)
```
This method can be removed when other changes are made.

rstam commented on src/MongoDB.Driver.Core/Core/ConnectionPools/ExclusiveConnectionPool.cs:
```diff
@@ -357,11 +367,21 @@ private void ReleaseConnection(PooledConnection connection)
             if (_state.Value == State.Disposed)
             {
-                throw new ObjectDisposedException(GetType().Name);
+                ThrowAndLogException(new ObjectDisposedException(GetType().Name));
```
Keep original (just throw).

rstam commented on src/MongoDB.Driver.Core/Core/ConnectionPools/ExclusiveConnectionPool.cs:
```diff
@@ -370,7 +390,7 @@ private void ThrowIfNotOpen()
             {
                 ThrowIfDisposed();
-                throw new InvalidOperationException("ConnectionPool must be initialized.");
+                ThrowAndLogException(new InvalidOperationException("ConnectionPool must be initialized."));
```
Keep original (just throw).

rstam commented on src/MongoDB.Driver.Core/Core/ConnectionPools/ExclusiveConnectionPool.cs:
```diff
@@ -394,17 +414,17 @@ private class AcquireConnectionHelper
 
+            // public methods
+            public void CheckingOutConnection()
+            {
```
Split `CheckingOutConnection` into two methods:

            // public methods
            public void CheckingOutConnection()
            {
                var handler = _pool._checkingOutConnectionEventHandler;
                if (handler != null)
                {
                    handler(new ConnectionPoolCheckingOutConnectionEvent(_pool._serverId, EventContext.OperationId));
                }
            }

            public void EnterWaitQueue()
            {
                _enteredWaitQueue = _pool._waitQueue.Wait(0); // don't wait...
                if (!_enteredWaitQueue)
                {
                    throw MongoWaitQueueFullException.ForConnectionPool(_pool._endPoint);
                }

                _stopwatch = Stopwatch.StartNew();
            }

DmitryLukyanov replied:
done

rstam commented on src/MongoDB.Driver.Core/Core/ConnectionPools/ExclusiveConnectionPool.cs:
```diff
@@ -467,7 +487,10 @@ public void HandleException(Exception ex)
-                    handler(new ConnectionPoolCheckingOutConnectionFailedEvent(_pool._serverId, ex, EventContext.OperationId));
+                    var reason = ex is TimeoutException
+                        ? ConnectionCloseReason.Timeout
+                        : ConnectionCloseReason.PoolClosed;
```
I think there are three cases:

    switch (ex)
    {
        case ObjectDisposedException _: reason = ConnectionCheckOutFailedReason.PoolClosed; break;
        case TimeoutException _: reason = ConnectionCheckOutFailedReason.Timeout; break;
        default: reason = ConnectionCheckOutFailedReason.ConnectionError; break;
    }

rstam replied:
Replace `Clear` method (far below, but no way to add comment there) with:

    public void Clear()
    {
        List<PooledConnection> connectionsToClose;
        lock (_lock)
        {
            connectionsToClose = new List<PooledConnection>(_connections);
            foreach (var connection in _connections)
            {
                RemoveConnection(connection);
            }
            _connections.Clear();
        }

        foreach (var connection in connectionsToClose)
        {
            connection.Dispose();
        }
    }

rstam replied:
Replace `Prune` method (far below, but no way to add comment there) with:

    public void Prune()
    {
        PooledConnection expiredConnection = null;
        lock (_lock)
        {
            for (int i = 0; i < _connections.Count; i++)
            {
                if (_connections[i].IsExpired)
                {
                    expiredConnection = _connections[i];
                    RemoveConnection(expiredConnection);
                    _connections.RemoveAt(i);
                    break;
                }
            }
        }

        if (expiredConnection != null)
        {
            expiredConnection.Dispose();
        }
    }

rstam replied:
In `Aquire` method (far below, but no way to add comment here), add:

    connection.Dispose();

after call to `RemoveConnection`.

rstam replied:
Replace `Return` method (far below, but no way to add comment there) with:

    public void Return(PooledConnection connection, out bool closeConnection)
    {
        closeConnection = false;
        if (connection.IsExpired)
        {
            RemoveConnection(connection);
            closeConnection = true;
            return;
        }

        lock (_lock)
        {
            _connections.Add(connection);
        }
    }

rstam replied:
In `RemoveConnection` method (far below, but no way to add comment there), remove:

    connection.Dispose();

rstam commented on src/MongoDB.Driver.Core/Core/Events/ConnectionCloseReason.cs:
```diff
@@ -0,0 +1,32 @@
+﻿/* Copyright 2019-present MongoDB Inc.
```
Add new enum:

    namespace MongoDB.Driver.Core.Events
    {
        /// <summary>
        /// Represents the reason an attempt to check out a connection failed.
        /// </summary>
        public enum ConnectionCheckOutFailedReason
        {
            /// <summary>
            /// The connection pool is closed.
            /// </summary>
            PoolClosed,
    
            /// <summary>
            /// Timeout waiting for a connection to become available.
            /// </summary>
            Timeout,
    
            /// <summary>
            /// Connection error while opening a new connection.
            /// </summary>
            ConnectionError
        }
    }

rstam commented on src/MongoDB.Driver.Core/Core/Events/ConnectionCloseReason.cs:
```diff
@@ -0,0 +1,32 @@
+    /// <summary>
+    /// Represents the connection close reasons.
+    /// </summary>
+    public enum ConnectionCloseReason
```
Rename to `ConnectionClosedReason` (add "d")

But this enum isn't actually used yet.

rstam commented on src/MongoDB.Driver.Core/Core/Events/ConnectionCloseReason.cs:
```diff
@@ -0,0 +1,32 @@
+        /// <summary>
+        /// Timeout exception.
+        /// </summary>
+        Timeout
```
The values are not correct. Replace with:

    /// <summary>
    /// Represents the reason a connection was closed.
    /// </summary>
    public enum ConnectionClosedReason
    {
        /// <summary>
        /// The pool was cleared, making the connection no longer valid.
        /// </summary>
        Stale,

        /// <summary>
        /// The connection became statle by being available for too long.
        /// </summary>
        Idle,

        /// <summary>
        /// The connection experienced an error, making it no longer valid.
        /// </summary>
        Error,

        /// <summary>
        /// The pool was closed, making the connection no longer valid.
        /// </summary>
        PoolClosed,

        /// <summary>
        /// The reason the connection was closed is unknown.
        /// </summary>
        Unknown
    }

rstam commented on src/MongoDB.Driver.Core/Core/Events/ConnectionPoolCheckingOutConnectionFailedEvent.cs:
```diff
@@ -35,11 +36,25 @@ public struct ConnectionPoolCheckingOutConnectionFailedEvent
+        /// </summary>
+        public ConnectionCloseReason? CloseReason
+        {
+            get { return _closeReason; }
```
We also need a new `ConnectionCreatedEvent`:

    /* Copyright 2019-present MongoDB Inc.
    *
    * Licensed under the Apache License, Version 2.0 (the "License");
    * you may not use this file except in compliance with the License.
    * You may obtain a copy of the License at
    *
    * http://www.apache.org/licenses/LICENSE-2.0
    *
    * Unless required by applicable law or agreed to in writing, software
    * distributed under the License is distributed on an "AS IS" BASIS,
    * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    * See the License for the specific language governing permissions and
    * limitations under the License.
    */

    using MongoDB.Driver.Core.Clusters;
    using MongoDB.Driver.Core.Configuration;
    using MongoDB.Driver.Core.Connections;
    using MongoDB.Driver.Core.Servers;

    namespace MongoDB.Driver.Core.Events
    {
        /// <preliminary/>
        /// <summary>
        /// Occurs when a connection is created..
        /// </summary>
        /// <remarks>Corresponds to the ConnectionCreatedEvent from the CMAP spec.</remarks>
        public struct ConnectionCreatedEvent
        {
            private readonly ConnectionId _connectionId;
            private readonly ConnectionSettings _connectionSettings;
            private readonly long? _operationId;

            /// <summary>
            /// Initializes a new instance of the <see cref="ConnectionCreatedEvent" /> struct.
            /// </summary>
            /// <param name="connectionId">The connection identifier.</param>
            /// <param name="connectionSettings">The connection settings.</param>
            /// <param name="operationId">The operation identifier.</param>
            public ConnectionCreatedEvent(ConnectionId connectionId, ConnectionSettings connectionSettings, long? operationId)
            {
                _connectionId = connectionId;
                _connectionSettings = connectionSettings;
                _operationId = operationId;
            }

            /// <summary>
            /// Gets the cluster identifier.
            /// </summary>
            public ClusterId ClusterId
            {
                get { return _connectionId.ServerId.ClusterId; }
            }

            /// <summary>
            /// Gets the connection identifier.
            /// </summary>
            public ConnectionId ConnectionId
            {
                get { return _connectionId; }
            }

            /// <summary>
            /// Gets the connection settings.
            /// </summary>
            public ConnectionSettings ConnectionSettings
            {
                get { return _connectionSettings; }
            }

            /// <summary>
            /// Gets the operation identifier.
            /// </summary>
            public long? OperationId
            {
                get { return _operationId; }
            }

            /// <summary>
            /// Gets the server identifier.
            /// </summary>
            public ServerId ServerId
            {
                get { return _connectionId.ServerId; }
            }
        }
    }

rstam commented on src/MongoDB.Driver.Core/Core/Events/ConnectionPoolCheckingOutConnectionFailedEvent.cs:
```diff
@@ -25,6 +25,7 @@ namespace MongoDB.Driver.Core.Events
     /// </summary>
     public struct ConnectionPoolCheckingOutConnectionFailedEvent
     {
+        private readonly ConnectionCloseReason? _closeReason;
```
Should be `ConnectionCheckOutFailedReason` and doesn't need to be nullable.

rstam commented on src/MongoDB.Driver.Core/Core/Events/ConnectionPoolCheckingOutConnectionFailedEvent.cs:
```diff
@@ -35,11 +36,25 @@ public struct ConnectionPoolCheckingOutConnectionFailedEvent
+        /// <summary>
+        /// Gets the close reason.
+        /// </summary>
+        public ConnectionCloseReason? CloseReason
```
Return type should be `ConnectionCheckOutFailedReason` and name can be just `Reason`

rstam commented on tests/MongoDB.Bson.TestHelpers/Reflector.cs:
```diff
@@ -139,6 +139,12 @@ public static void SetFieldValue(object obj, string name, object value, BindingF
             fieldInfo.SetValue(obj, value);
         }
 
+        public static void SetStaticField(Type type, string name, object value)
```
Rename to `SetStaticFieldValue`

rstam commented on tests/MongoDB.Driver.Core.TestHelpers/EventCapturer.cs:
```diff
@@ -98,6 +100,12 @@ public bool Any()
+        public Task NotifyWhen(Func<IEnumerable<object>, bool> condition)
+        {
+            var notifier = new EventNotifier(this, condition);
+            return notifier.TaskCompletion;
```
Replace with:

    return notifier.Initialize();

DmitryLukyanov replied:
done

rstam commented on tests/MongoDB.Driver.Core.TestHelpers/EventCapturer.cs:
```diff
@@ -244,5 +254,40 @@ private BsonJavaScriptWithScope RecursivelyMaterialize(BsonJavaScriptWithScope v
+
+        private class EventNotifier
+        {
+            private readonly Func<IEnumerable<object>, bool> _condition;
```
Add:

    private readonly EventCapturer _eventCapturer;

DmitryLukyanov replied:
done

rstam commented on tests/MongoDB.Driver.Core.TestHelpers/EventCapturer.cs:
```diff
@@ -244,5 +254,40 @@ private BsonJavaScriptWithScope RecursivelyMaterialize(BsonJavaScriptWithScope v
+                Ensure.IsNotNull(eventCapturer, nameof(eventCapturer));
+                _taskCompletionSource = new TaskCompletionSource<bool>();
+                _condition = Ensure.IsNotNull(condition, nameof(condition));
+                InitializeTask(eventCapturer);
```
I don't like calling `Initialize` from the constructor. Let's have the creator of the object call `Initialize`.

DmitryLukyanov replied:
done

rstam commented on tests/MongoDB.Driver.Core.TestHelpers/EventCapturer.cs:
```diff
@@ -244,5 +254,40 @@ private BsonJavaScriptWithScope RecursivelyMaterialize(BsonJavaScriptWithScope v
+
+            public EventNotifier(EventCapturer eventCapturer, Func<IEnumerable<object>, bool> condition)
+            {
+                Ensure.IsNotNull(eventCapturer, nameof(eventCapturer));
```
Replace with:

    _eventCapturer = Ensure.IsNotNull(eventCapturer, nameof(eventCapturer));

DmitryLukyanov replied:
done

rstam commented on tests/MongoDB.Driver.Core.TestHelpers/EventCapturer.cs:
```diff
@@ -244,5 +254,40 @@ private BsonJavaScriptWithScope RecursivelyMaterialize(BsonJavaScriptWithScope v
+            public Task TaskCompletion => _taskCompletionSource.Task;
+
+            // private methods
+            private void InitializeTask(EventCapturer eventCapturer)
```
Make public and rename to `Initialize`:

    // public methods
    public Task Initialize()
    {
        _eventCapturer._addEventAction += (events, @event) => TriggerNotificationIfCondition(events);

        // condition might already be true even before any more events are added
        lock (_eventCapturer._lock)
        {
            TriggerNotificationIfCondition(_eventCapturer.Events);
        }

        return _taskCompletionSource.Task;
    }

DmitryLukyanov replied:
done

rstam commented on tests/MongoDB.Driver.Core.Tests/Specifications/connection-monitoring-and-pooling/ConnectionMonitoringAndPoolingTestRunner.cs:
```diff
@@ -0,0 +1,607 @@
+            nameof(ConnectionPoolRemovedConnectionEvent),
+            nameof(ConnectionPoolClosingEvent),
+            nameof(ConnectionPoolClearingEvent),
+            nameof(ConnectionClosingEvent)
```
I'm suggesting slightly different mappings between the spec CMAP events and driver events. Specifically I do not want to map any driver `-ing` events (which are raised *before* something happens) to CMAP `-ed` events (which are intended to be raised *after* something has happened).

Therefore the list of always ignored events is slightly different:

    private static readonly string[] __alwaysIgnoredEvents =
    {
        nameof(ConnectionPoolOpeningEvent),
        nameof(ConnectionPoolAddingConnectionEvent),
        nameof(ConnectionPoolAddedConnectionEvent),
        nameof(ConnectionPoolCheckingInConnectionEvent),
        nameof(ConnectionPoolRemovingConnectionEvent),
        nameof(ConnectionPoolRemovedConnectionEvent),
        nameof(ConnectionPoolClosingEvent),
        nameof(ConnectionPoolClearingEvent),
        nameof(ConnectionOpeningEvent),
        nameof(ConnectionClosingEvent)
    };

DmitryLukyanov replied:
done

rstam commented on tests/MongoDB.Driver.Core.Tests/Specifications/connection-monitoring-and-pooling/ConnectionMonitoringAndPoolingTestRunner.cs:
```diff
@@ -0,0 +1,607 @@
+                {
+                    actualEvent.CloseReason().Should().NotBeNull();
+                    actualEvent.CloseReason().ToString().ToLower().Should().Be(reason.ToString().ToLower());
+                }
```
Replace lines 176-180 with:

    if (actualEvent is ConnectionPoolCheckingOutConnectionFailedEvent checkingOutFailedEvent)
    {
        var actualReason = checkingOutFailedEvent.Reason;
        actualReason.Should().NotBeNull();
        actualReason.ToString().ToLower().Should().Be(reason.ToString().ToLower());
    }
    else if (actualEvent is ConnectionClosedEvent connectionClosedEvent)
    {
        // we don't support Reason for ConnectionClosedEvent
    }
    else
    {
        throw new Exception($"Don't know how to assert against reason for class : {actualEvent.GetType().FullName}.");
    }

DmitryLukyanov replied:
done

rstam commented on tests/MongoDB.Driver.Core.Tests/Specifications/connection-monitoring-and-pooling/ConnectionMonitoringAndPoolingTestRunner.cs:
```diff
@@ -0,0 +1,607 @@
+            }
+        }
+
+        private string MapExpectedEventName(string expectedEventName)
```
Rename to `MapExpectedEventNameToDriverTypeName`

rstam commented on tests/MongoDB.Driver.Core.Tests/Specifications/connection-monitoring-and-pooling/ConnectionMonitoringAndPoolingTestRunner.cs:
```diff
@@ -0,0 +1,607 @@
+            switch (expectedEventName)
+            {
+                case "ConnectionPoolCreated":
+                    return nameof(ConnectionPoolOpeningEvent);
```
return nameof(ConnectionPoolOpenedEvent);

`-ed` not `-ing`

rstam commented on tests/MongoDB.Driver.Core.Tests/Specifications/connection-monitoring-and-pooling/ConnectionMonitoringAndPoolingTestRunner.cs:
```diff
@@ -0,0 +1,607 @@
+                case "ConnectionPoolCleared":
+                    return nameof(ConnectionPoolClearedEvent);
+                case "ConnectionReady":
+                    return null;
```
return nameof(ConnectionOpenedEvent);

rstam commented on tests/MongoDB.Driver.Core.Tests/Specifications/connection-monitoring-and-pooling/ConnectionMonitoringAndPoolingTestRunner.cs:
```diff
@@ -0,0 +1,607 @@
+                case "ConnectionReady":
+                    return null;
+                case "ConnectionCheckedIn":
+                    return nameof(ConnectionPoolCheckingInConnectionEvent);
```
return nameof(ConnectionPoolCheckedInConnectionEvent);

`-ed` not `-ing`

rstam commented on tests/MongoDB.Driver.Core.Tests/Specifications/connection-monitoring-and-pooling/ConnectionMonitoringAndPoolingTestRunner.cs:
```diff
@@ -0,0 +1,607 @@
+                case "ConnectionClosed":
+                    return nameof(ConnectionClosedEvent);
+                case "ConnectionCreated":
+                    return nameof(ConnectionPoolAddedConnectionEvent);
```
return nameof(ConnectionCreatedEvent);

rstam commented on tests/MongoDB.Driver.Core.Tests/Specifications/connection-monitoring-and-pooling/ConnectionMonitoringAndPoolingTestRunner.cs:
```diff
@@ -0,0 +1,607 @@
+            var expectedCount = operation.GetValue("count").ToInt32();
+            var notifyTask = eventCapturer.NotifyWhen(coll => coll.Count(c => c.GetType().Name == eventType) >= expectedCount);
+
+            var testFailedTimeout = Task.Delay(TimeSpan.FromMinutes(5), CancellationToken.None);
```
I changed the timeout to 1 minute to fail quicker while testing.

DmitryLukyanov replied:
I didn't know which timeout is better. 1 min is fine for me too.

rstam commented on tests/MongoDB.Driver.Core.Tests/Specifications/connection-monitoring-and-pooling/ConnectionMonitoringAndPoolingTestRunner.cs:
```diff
@@ -0,0 +1,607 @@
+                var version = document.GetValue("version", null)?.ToInt32() ?? 1;
+                foreach (var async in new[] { false, true })
+                {
+                    var name = GetTestCaseName(document, document, version);
```
Third argument should be `index` not `version`:

    var name = GetTestCaseName(document, document, index);

rstam commented on tests/MongoDB.Driver.Core.Tests/Specifications/connection-monitoring-and-pooling/ConnectionMonitoringAndPoolingTestRunner.cs:
```diff
@@ -0,0 +1,607 @@
+
+            protected override IEnumerable<JsonDrivenTestCase> CreateTestCases(BsonDocument document)
+            {
+                var version = document.GetValue("version", null)?.ToInt32() ?? 1;
```
`version` is not actually needed here, but `index` is, so replace with:

    var index = 0;

Could be `-1` also since we don't really know the index.

rstam commented on tests/MongoDB.Driver.Core.Tests/Specifications/connection-monitoring-and-pooling/ConnectionMonitoringAndPoolingTestRunner.cs:
```diff
@@ -0,0 +1,607 @@
+
+    internal static class CmapEventsReflector
+    {
+        public static ConnectionCloseReason? CloseReason(this object @event)
```
Remove this method (see changes that make it unnecessary).

rstam commented on tests/MongoDB.Driver.Core.Tests/Specifications/connection-monitoring-and-pooling/ConnectionMonitoringAndPoolingTestRunner.cs:
```diff
@@ -0,0 +1,607 @@
+            return (ConnectionCloseReason?)Reflector.GetPropertyValue(@event, nameof(CloseReason), BindingFlags.Public | BindingFlags.Instance);
+        }
+
+        public static ConnectionId ConnectionId(this object @event)
```
I *really* don't like extension methods on `object` because they are extremely dangerous.

For example, you could write:

    42.ConnectionId()

and it will compile fine even though it makes no sense.

I understand that you want to use this same reflector method with multiple event types and that there is no other common ancestor besides `object`, and since this extension method is somewhat buried inside this particular namespace we could let it slide.

rstam commented on src/MongoDB.Driver.Core/Core/ConnectionPools/ExclusiveConnectionPool.cs:
```diff
@@ -394,17 +414,17 @@ private class AcquireConnectionHelper
                 if (handler != null)
                 {
                     handler(new ConnectionPoolCheckingOutConnectionEvent(_pool._serverId, EventContext.OperationId));
                 }
```
Let's NOT raise this event from the constructor. Let's leave it in `CheckingOutConnection` like it was.

DmitryLukyanov replied:
done

rstam commented on tests/MongoDB.Driver.Core.Tests/Specifications/connection-monitoring-and-pooling/ConnectionMonitoringAndPoolingTestRunner.cs:
```diff
@@ -0,0 +1,607 @@
+                action,
+                CancellationToken.None,
+                TaskCreationOptions.None,
+                new ThreadPerTaskScheduler());
```
I really don't understand why `ThreadPerTaskScheduler` is needed, but I also don't see that it causes any harm.

rstam replied:
I changed this to:

    return Task.Factory.StartNew(
        action,
        CancellationToken.None);

and all tests still pass.

I don't think there is any harm in using `ThreadPerTaskScheduler` but I still don't understand why you think it is necessary.

DmitryLukyanov replied:
Discussed offline. Leaving it as is, but additionally, I've added a comment with an explanation of the idea of this task

rstam replied:
I remain unconvinced that a custom thread scheduler is needed, so the remaining question is does leaving it cause any harm. I can answer either way:

1. It causes no harm because the tests still pass
2. It causes some conceptual harm because if it's not needed why is it there

I understand of course what the custom scheduler does, I just don't see why it is necessary for adequate test coverage.

The end result though is that Dmitry is adding a comment that we're not sure this custom scheduler is necessary but we're keeping it just in case.

rstam commented:
The changes I am suggesting have been prototyped here:

https://github.com/rstam/mongo-csharp-driver/tree/csharp2427-robert

DmitryLukyanov reviewed (commented):
Some comments still under discussion.
The current implementation has fewer changes and looks more similar to the previously implemented.

rstam commented:
I have found a simpler refactoring that raises the events in the right order. See:

https://github.com/rstam/mongo-csharp-driver/tree/csharp2427-robert-20191009

Notes:

1. `ListConnectionHolder.Return` no longer checks whether the connection is expired, it simply returns the connection to the pool. The caller decides whether to return the connection to the pool or remove it.

2. `ListConnectionHolder.RemoveConnection` is now public so it can be called. It continues to call `connection.Dispose()` so your concern about splitting `RemoveConnection` and `Dispose` are addressed

3. The remaining refactoring is in `ExclusiveConnectionPool.ReleaseConnection` which now has a much more straightforward and understandable execution path (i.e. it raises a few events and either returns the connection to the pool or removes it)

DmitryLukyanov commented:
Ready for review.

rstam commented on src/MongoDB.Driver.Core/Core/Configuration/ConnectionPoolSettings.cs:
```diff
@@ -93,6 +93,7 @@ public int MinConnections
         /// <value>
         /// The size of the wait queue.
         /// </value>
+        [Obsolete("This functionality will be replaced in next releases.")]
```
Let's say "property" instead of "functionality" and "a later release" instead of "next releases".

    [Obsolete("This property will be removed in a later release.")]

"property" is more accurate and it will only be removed once so "a later release" is singular instead of plural.

Find all other instances and make the same change.

DmitryLukyanov replied:
done

rstam commented on src/MongoDB.Driver.Core/Core/Events/ConnectionClosedReason.cs:
```diff
@@ -0,0 +1,51 @@
+{
+    /// <summary>
+    /// Represents the connection close reasons.
+    /// </summary>
```
Remove the first summary and keep the second.

DmitryLukyanov replied:
done

rstam commented on src/MongoDB.Driver.Core/Core/Events/ConnectionClosedReason.cs:
```diff
@@ -0,0 +1,51 @@
+        Stale,
+
+        /// <summary>
+        /// The connection became statle by being available for too long.
```
`stale` (not `statle`)

DmitryLukyanov replied:
done

rstam commented on src/MongoDB.Driver.Core/Core/Events/ConnectionPoolCheckingOutConnectionFailedEvent.cs:
```diff
@@ -35,11 +36,25 @@ public struct ConnectionPoolCheckingOutConnectionFailedEvent
+        /// <summary>
+        /// Gets the reason the checkout failed.
+        /// </summary>
+        public ConnectionCheckOutFailedReason Reason
```
Put in alphabetical order?

DmitryLukyanov replied:
done

rstam commented on src/MongoDB.Driver.Core/Core/Events/ConnectionPoolClearedEvent.cs:
```diff
@@ -0,0 +1,66 @@
+    public struct ConnectionPoolClearedEvent
+    {
+        private readonly ServerId _serverId;
+        private readonly ConnectionPoolSettings _connectionPoolSettings;
```
Put in alphabetical order?

DmitryLukyanov replied:
done

rstam commented on src/MongoDB.Driver.Core/Core/Events/ConnectionPoolClearingEvent.cs:
```diff
@@ -0,0 +1,66 @@
+    public struct ConnectionPoolClearingEvent
+    {
+        private readonly ServerId _serverId;
+        private readonly ConnectionPoolSettings _connectionPoolSettings;
```
Put in alphabetical order?

DmitryLukyanov replied:
done

rstam commented on src/MongoDB.Driver/MongoUrlBuilder.cs:
```diff
@@ -747,6 +753,7 @@ public void Parse(string url)
             else if (connectionString.WaitQueueMultiple != null)
             {
                 _waitQueueMultiple = connectionString.WaitQueueMultiple.Value;
+#pragma warning restore 618
```
It seems a little unbalanced to restore the warning in the middle of the if statement.

Should we try to use #pragmas around whole statements (even if they are compound statements like if)?

If so that would mean moving the restore to after line 758.

DmitryLukyanov replied:
done

rstam commented on tests/MongoDB.Driver.Core.Tests/Specifications/connection-monitoring-and-pooling/ConnectionMonitoringAndPoolingTestRunner.cs:
```diff
@@ -0,0 +1,613 @@
+        public void RunTestDefinition(JsonDrivenTestCase testCase)
+        {
+            var connectionMap = new ConcurrentDictionary<string, IConnection>();
+            var eventCapturer = new Lazy<EventCapturer>(isThreadSafe: true);
```
I can't think of any reason this needs to be `Lazy`.

I removed the `Lazy` and tests still pass.

Am I missing something?

DmitryLukyanov replied:
Discussed offline. My original implementation was based on a misunderstanding of what a `isThreadSafe: true` flag does.
Done.

rstam commented on tests/MongoDB.Driver.Core.Tests/Specifications/connection-monitoring-and-pooling/ConnectionMonitoringAndPoolingTestRunner.cs:
```diff
@@ -0,0 +1,613 @@
+            EnsureAvailableStyle(test);
+            ResetConnectionId();
+
+            var connectionPool = new Lazy<IConnectionPool>(
```
I can't think of any reason this needs to be `Lazy`.

I removed the `Lazy` and tests still pass.

Am I missing something?

DmitryLukyanov replied:
The same as above

rstam commented on tests/MongoDB.Driver.Core.Tests/Specifications/connection-monitoring-and-pooling/ConnectionMonitoringAndPoolingTestRunner.cs:
```diff
@@ -0,0 +1,613 @@
+                    tasks,
+                    connectionPool.Value,
+                    async,
+                    out exception);
```
Would it make sense to have a single try/catch at this level so each operation doesn't have to have its own try/catch to set the exception variable?

DmitryLukyanov replied:
I guess the current way is better. With your suggestion, we will need to filter the throwing of our own errors like in this line: https://github.com/DmitryLukyanov/mongo-csharp-driver/pull/57/files#diff-a638783da40724c2c11679e9321e1829R268 and etc

rstam replied:
Why would we need to filter our own exceptions?

If an exception was thrown either it matches the expected exception or it doesn't.

rstam replied:
But you can use multiple try/catch if you wish. It's not stopping me from aprroving the PR.

DmitryLukyanov replied:
The idea of the `AssertError` method (at least for me) was to validate errors from the tested driver code, not from the tests itself. So, in case if we see (for example in the evergreen results) an error which contains an error like: 

    Unexpected exception has been thrown

https://github.com/DmitryLukyanov/mongo-csharp-driver/pull/57/files#diff-a638783da40724c2c11679e9321e1829R101

It will mean that something wrong in the driver code only. And it's not related to, for example, the fact that JSON files were changed.
It's not a big deal to implement your suggestion, but I would prefer to leave it as is.

However. I noticed one minor problem when we could overwrite the exception if it has been thrown more than in one step. Fixed in the last commit.

rstam reviewed (changes requested):
It's looking very good. Only minor comments left to address.

DmitryLukyanov commented:
ready for review.

DmitryLukyanov commented:
Evergreen: https://evergreen.mongodb.com/version/5d9f633f0305b9334093a124

rstam commented on src/MongoDB.Driver.Core/Core/Events/ConnectionClosedReason.cs:
```diff
@@ -0,0 +1,48 @@
+        Stale,
+
+        /// <summary>
+        /// The connection became stale  by being available for too long.
```
Probably doesn't matter but there appear to be two spaces after the word "stale".

DmitryLukyanov replied:
done

rstam commented on src/MongoDB.Driver.Core/Core/Events/ConnectionCreatedEvent.cs:
```diff
@@ -0,0 +1,87 @@
+
+namespace MongoDB.Driver.Core.Events
+{
+    /// <preliminary/>
```
I realize this was copy/pasted from another event... but I just noticed it.

I didn't remember that `<preliminary>` even existed... and now that I see it wonder whether it belongs here. After all this event is from a spec... it's probably not so preliminary any more.

And what about removing `<preliminary>` from other events, either in this ticket or in a separate ticket.

DmitryLukyanov replied:
I will remove all these tags

rstam commented on tests/MongoDB.Driver.Core.Tests/Specifications/connection-monitoring-and-pooling/ConnectionMonitoringAndPoolingTestRunner.cs:
```diff
@@ -0,0 +1,609 @@
+            }
+
+            AssertError(test, exception);
+            AssertEvents(test, eventCapturer.Value);
```
This line does not compile.

DmitryLukyanov replied:
Fixed

DmitryLukyanov commented:
ready for review

DmitryLukyanov commented:
Evergreen:  https://evergreen.mongodb.com/version/5d9f729f3e8e8617cc6893a7

rstam commented on src/MongoDB.Driver.Core/Core/Events/ConnectionCreatedEvent.cs:
```diff
@@ -0,0 +1,87 @@
+    /// <summary>
+    /// Occurs when a connection is created..
+    /// </summary>
+    /// <remarks>Corresponds to the ConnectionCreatedEvent from the CMAP spec.</remarks>
```
None of the other events have a comment like this.

Should we remove this <remarks> line?

DmitryLukyanov replied:
done

rstam reviewed (approved):
One minor optional change.

LGTM

vincentkam commented on src/MongoDB.Driver.Core/Core/Events/ConnectionPoolClearedEvent.cs:
```diff
@@ -0,0 +1,65 @@
+namespace MongoDB.Driver.Core.Events
+{
+    /// <summary>
+    /// Occurs after the pool is opened.
```
opened -> cleared ?

DmitryLukyanov replied:
Fixed. Will be pushed with all other potential changes

vincentkam commented on src/MongoDB.Driver.Core/Core/Events/ConnectionPoolClearingEvent.cs:
```diff
@@ -0,0 +1,65 @@
+namespace MongoDB.Driver.Core.Events
+{
+    /// <summary>
+    /// Occurs after the pool is opened.
```
`Occurs when the pool is about to be cleared` ?

DmitryLukyanov replied:
Fixed. Will be pushed with all other potential changes

DmitryLukyanov commented:
Evergreen: https://evergreen.mongodb.com/version/5d9fac593e8e8617cc6cca9c
Fixed the previous unit tests.

DmitryLukyanov commented:
Added the rest of the required changes regarding CSHARP-2664.

Also, see these prose tests: https://github.com/mongodb/specifications/blob/ef782f2be3bfa32cb3c2befb63d85de142df31aa/source/connection-monitoring-and-pooling/tests/README.rst#prose-tests

1.2. I don't think that we need to implement the first 2 tests. Since they are too common.
3. The 3rd is partially implemented in the ConnectionStringTests. For example here: https://github.com/mongodb/mongo-csharp-driver/blob/master/tests/MongoDB.Driver.Core.Tests/Core/Configuration/ConnectionStringTests.cs#L585

I think it's enough to have only this test(and similar in this file) because we validate that this option for `ConnectionPool`(only) is presented in a connectionString. I don't think that there is a big chance that this option can be used for any other target except connection pool settings. So, I guess this prose test can be considered as implemented.
4. The target for this test is a base functionality on which the CMAP test runner(and a lot of other tests) has been built. So, I guess we should not worry about this test at all.
5. I've updated some tests to validate generated events on a failed reason(not only on `connectionError`) in the last commit

rstam commented:
I thought we were done with this PR.

Are you starting work on a new ticket using the same PR?

DmitryLukyanov commented:
>Are you starting work on a new ticket using the same PR?

we've almost implemented the required changes for the ticket 2664. The only changes which I've added are about updating already created tests. It's several lines change, so I don't think that a new PR is required for it.

rstam commented:
> > Are you starting work on a new ticket using the same PR?
> 
> we've almost implemented the required changes for the ticket 2664. The only changes which I've added are about updating already created tests. It's several lines change, so I don't think that a new PR is required for it.

OK. I had not yet seen the discussion about this on Slack when I asked the question here.

Thanks.

rstam commented on tests/MongoDB.Driver.Core.Tests/Core/ConnectionPools/ExclusiveConnectionPoolTests.cs:
```diff
@@ -117,6 +128,11 @@ public void Constructor_should_throw_when_eventSubscriber_is_null()
             }
 
             act.ShouldThrow<InvalidOperationException>();
+            _capturedEvents.Next(); // skip ConnectionPoolCheckingOutConnectionEvent
```
It might be a good idea to assert that the event we are skipping is in fact of the expected type:

    _capturedEvents.Next().Should().BeOfType<ConnectionPoolCheckingOutConnectionEvent>(); // skip ConnectionPoolCheckingOutConnectionEvent

rstam replied:
Same for other places where you've added code to skip events.

DmitryLukyanov replied:
done

rstam reviewed (approved):
LGTM with minor comments.

DmitryLukyanov commented:
Evergreen: https://evergreen.mongodb.com/version/5da084b657e85a640d6112c9

vincentkam reviewed (changes requested):
Partial review (still reviewing), but posting my what feedback I do have for now as requested by Dima :)

vincentkam commented on tests/MongoDB.Driver.Core.Tests/Core/Helpers/MockConnection.cs:
```diff
@@ -88,10 +142,18 @@ public List<RequestMessage> GetSentMessages()
         {
+            _openedAtUtc = DateTime.UtcNow;
+            // _lastUsedAtUtc is set in the SendBuffer method in the BinaryConnection
+            // which is one from methods called inside Open
```
Did you mean "which is one of the methods called from inside Open"?

DmitryLukyanov replied:
yep, I will fix it

vincentkam commented on tests/MongoDB.Driver.Core.Tests/Core/Helpers/MockConnection.cs:
```diff
@@ -88,10 +142,18 @@ public List<RequestMessage> GetSentMessages()
         {
+            _openedAtUtc = DateTime.UtcNow;
+            // _lastUsedAtUtc is set in the SendBufferAsync method in the BinaryConnection
+            // which is one from methods called inside OpenAsync
```
Did you mean "which is one of the methods called from inside OpenAsync"?

DmitryLukyanov replied:
yep, I will fix it

vincentkam commented on src/MongoDB.Driver.Core/Core/ConnectionPools/ExclusiveConnectionPool.cs:
```diff
@@ -140,12 +148,12 @@ public int UsedCount
             try
             {
                 helper.CheckingOutConnection();
+                ThrowIfNotOpen();
```
Just to confirm:
The "old" sequence of execution was essentially: ThrowIfDisposed, EnterWaitQueue, emit "CheckingOutConnection" event.
The "new" sequence of execution is: Emit "CheckingOutConnection" event, ThrowIfDisposed, EnterWaitQueue, 

Would changing this sequence of execution potentially be considered a breaking API change? Is there any scenario in which an application could rely on the old behavior?
@rstam

rstam replied:
So it's not a breaking change in the sense of your application no longer compiles.

But it is a slight change in behavior which may or may not be breaking.

However, we don't really have a choice. The spec (the JSON test files specifically) mandate that the events occur in a certain order.

Furthermore, I think the order in the JSON test files is more logical than the order we were raising the events in, otherwise we might have pushed for a spec change.

vincentkam replied:
Agreed, we have to make this change. I was merely wondering with what version we'd release this change. After some additional research, I think the order in which we raise events/throw exceptions is an implementation detail and since we have no evidence to believe that changing this order would break a large # of users, I'm fine with releasing this in a patch release.  Resolving this thread!

vincentkam commented:
(Above review is incomplete, but I was submitting early feedback per Dima's request so that we can better parallelize)

vincentkam commented on tests/MongoDB.Driver.Core.Tests/Specifications/connection-monitoring-and-pooling/ConnectionMonitoringAndPoolingTestRunner.cs:
```diff
@@ -0,0 +1,614 @@
+
+            if (actualEvents.Count > expectedEvents.Count)
+            {
+                throw new Exception($"Unexpected event of type: {actualEvents[expectedEvents.Count].GetType().Name}.");
```
I like this :)

vincentkam commented on tests/MongoDB.Driver.Core.Tests/Specifications/connection-monitoring-and-pooling/ConnectionMonitoringAndPoolingTestRunner.cs:
```diff
@@ -0,0 +1,614 @@
+        public void RunTestDefinition(JsonDrivenTestCase testCase)
+        {
+            var connectionMap = new ConcurrentDictionary<string, IConnection>();
+            var eventCapturer = new EventCapturer();
```
Capturing all the events by default will definitely help with debugging! 😸

vincentkam reviewed (approved):
Sorry, that took me longer  than I thought it would. LGTM!  😸

DmitryLukyanov committed:
CSHARP-2427: Implement Connection Monitoring and Pooling.
